### ISSUE TITLE
503 errors not handled

### ISSUE THREAD

**Issue by hamishfagg** [sondree/b2_fuse#16]
Hi,

According to Backblaze it is common to receive a 503 response to a request if a storage pod is busy.
From [this page](https://www.backblaze.com/b2/docs/uploading.html) on uploading:

```
// Try several times to upload the file.  It's normal
// for uploads to fail if the target storage pod is
// too busy.  It's also normal (but infrequent) to get
// a 429 Too Many Requests if you are uploading a LOT
// of files.
```

Here is the output from b2_fuse on such a response:

```
Headers: {'Content-Type': 'b2/x-auto', 'X-Bz-File-Name': 'data/0/839', 'X-Bz-Content-Sha1': '02894f9352d498d39517c01c2557cd8555a0bb4d', 'Content-Length': '13
99322', 'Authorization': '***********************************************************'}

{
  "code": "service_unavailable",
  "message": "c001_v0001001_t0037 is too busy",
  "status": 503
}
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/fuse.py", line 495, in _wrapper
    return func(*args, **kwargs) or 0
  File "/usr/local/lib/python2.7/dist-packages/fuse.py", line 621, in flush
    return self.operations('flush', self._decode_optional_path(path), fh)
  File "/usr/local/lib/python2.7/dist-packages/fuse.py", line 800, in __call__
    return getattr(self, op)(*args)
  File "b2fuse.py", line 456, in flush
    self.bucket.put_file(path, self.open_files[path])
  File "/home/mystx/b2_fuse/b2bucket_cached.py", line 149, in put_file
  File "/home/mystx/b2_fuse/b2bucket.py", line 148, in _put_file
  File "/home/mystx/b2_fuse/b2_python_pusher.py", line 71, in __enter__
SystemExit: 1
```

This causes my backup program to stop. Is there a way to get b2_fuse to retry?

Thanks

**Comment by falense** [comment]
At the moment there is no retry on server errors. Better error handling is
on the todo list, but will require significant restructuring.

Best

**Comment by hamishfagg** [comment]
Right. Will b2local fail in the same way?

**Comment by falense** [comment]
Yes

**Comment by ppolewicz** [comment]
Hi @sondree, b2 CLI developer here. When you started this project, b2 CLI was glued together with a console tool, which used `print` and `sys.exit(1)` if case of any error. Therefore it was unusable for the purpose of b2_fuse and you implemented the communication with B2 servers by yourself.

Since then, b2 CLI has changed a lot. Now the console tool is clearly separated from the API layer which is specifically designed for being used in other python programs such as filesystems.

Using b2 CLI API interface would give you several advantages over a custom built implementation of the communication layer:
- transient errors such as 503, 429, timeouts etc are retried automatically with the exponential backoff required by the recent interface change
- if the remote api changes, b2 CLI team will adapt the python API interface, so you'd just need to change the version of the CLI that b2_fuse depends on
- we use large file upload api when needed, we use threads etc

**Comment by falense** [comment]
@ppolewicz This is a really good suggestion. I will look into it.

**Comment by ppolewicz** [comment]
Encryption support is in the works too. If you'd use b2 cli API, fuse would be compatible with b2 cli buckets and will also feature encryption just by passing the passphrase into Bucket constructor (as soon as  we finish encryption).

**Comment by falense** [comment]
@ppolewicz I have had a look at B2 CLI, you guys have really come a long
way. I think it would be relatively straightforward to remap calls the
B2Fuse uses to your bucket API. This does appear like the best way moving
forward, not reinventing the wheel.

A couple of quick questions:

Does B2 CLI do any form of caching?

Do you do anything to improve performance for very large buckets (100k
files)? Getting directory structure for these buckets used to be a major
bottleneck.

From what I can tell, all calls to B2 CLI bucket are asynchronous. For some
use cases of B2 Fuse I need synchronous feedback, what would be the best
way of doing this?

Best
Sondre

2016-09-17 1:39 GMT+02:00 Paweł Polewicz notifications@github.com:

**Comment by ppolewicz** [comment]
Regarding caching, see [this](https://github.com/Backblaze/B2_Command_Line_Tool/blob/master/b2/cache.py). You can pass a caching object to `B2Api` constructor - if you won't, then `B2Api` will procure a `DummyCache` object and use that. I think `InMemoryCache` might be appropriate for your usage - I wrote it with fuse in mind (we test it, but we don't use it in `ConsoleTool`).

Currently cache does not store b2 directory structure, nor does a bucket object expect the cache object to do so. The thing is that if the user uses the same bucket from multiple processes, such case would become stale with first operation on the second thread. BUT! We can change it so that Bucket uses cache, but our `AccountInfoCache` has a very dummy cache for directory entries. Better this way than for you to inherit `Bucket` and do that by yourself. Please confirm that this is what you need and I'll code it.

I do not understand the synchronous/asynchronous issue. I think all API methods of `B2Api` and `Bucket` are purely synchronous, except the ones that return synchronously, but accept a `reporter` object, which receives updates during operation. We use that to show progress bars of upload/download/sync operations. Does that fit your need? If not, what would need to change?

**Comment by falense** [comment]
Actually, in order to get even decent performance B2 Fuse caches most calls
including directory listings. We can guarantee consistency assuming two
things: no concurrent access and reset of cache during bucket update calls.

For version 1.3 there is two major changes in planning:
1. Support for non-memory backed temporary storage. This is very useful
   working with large files.
2. Streaming of files in bucket.

When working on #1 I realized that one of the main issues currently with B2
Fuse and a large bucket (in addition to memory usage) is that you cannot
actually only get parts of the file you wish to read. This would greatly
alleviate the load of downloading large files when don't need to read the
entire file. This is a quite complicated feature (which is why it is taking
long), and does require background download of data in order to not leave
the user waiting for downloads. Hence some use cases of B2 Fuse will
require Asynchronous calls.

You should be able to try these two out:
https://github.com/sondree/b2_fuse/tree/release1.3-beta

I got in memory sparse files working along with asynchronous downloads,
which enables streaming. Next up is file backed temporary data storage.

On a side note v1.3 is moving away from the "local copy" version of B2 Fuse
for three reasons: not many seems to be using it, B2 CLI does this with
synchronization of folders and performance/features will be comparable to
file backed online only copy.

Best

2016-09-19 23:04 GMT+02:00 Paweł Polewicz notifications@github.com:

**Comment by ppolewicz** [comment]
I have read your bucket and threaded bucket classes.

I think we may be approaching this in a wrong way.

You should be our (b2 python api) client. Therefore you can ask us to provide an interface which abstracts stuff the filesystem should not care about (such as retrying, handling all kinds of tokens etc). One of the things that I think you request (and we don't yet support) is to be able to download a fragment of the file. It is easy. Just create a feature request in our tracker :) 

As for doing things in threads, we tried to avoid this as the user might prefer to use greenlets which would clash with our threads. There is a way around this and we may take it.

I can imagine sister classes for Bucket and B2Api, which return a future instead of a synchronous reply. Would that be a convenient design for you? 

As for caching, I am still not clear what you would expect. Can you pass us your caching object and we would delegate the work to it?

**Comment by ppolewicz** [comment]
I have read the caching classes too. I think this is something that most B2 Api users would not use, therefore lets maybe delegate it to the cache object that will be basically the cache that you are working on? Would that work?

**Comment by falense** [comment]
Sorry I haven't had time to reply.

Yes, that would work. I could also just convert the existing caching
structure to use your backend instead of mine.

Do you have an idea of how much work it would be for you guys to implement
a download into memory with optional fragment support? As far as I can tell
you dont support this today (only download to file).

I will have a closer look and a go at a first draft during the weekend.

Best
Sondre

**Comment by ppolewicz** [comment]
We already have an in-memory download destination [here](https://github.com/Backblaze/B2_Command_Line_Tool/blob/d96795edc755cf0b0f73864e2da52bbb5373d610/b2/download_dest.py#L112). You could implement our download destination interface (in such case please inherit from [`AbstractDownloadDestination`](https://github.com/Backblaze/B2_Command_Line_Tool/blob/d96795edc755cf0b0f73864e2da52bbb5373d610/b2/download_dest.py#L20)) or wrap our `DownloadDestBytes` object with your sparse caching layer.

What do you mean by optional fragment support?

**Comment by falense** [comment]
Just that it would be useful to be able to download parts of a file.
Optional refering to also being able to download whole files into memory.
As discussed previously, B2 CLI API does not support part of file downloads

**Comment by ppolewicz** [comment]
Ok, so since we have a memory destination already, we just need an optional `range` argument, which would download specific bytes from the target and return it. This is not a big thing to do. Judging by the amount of work you have to convert backends, I estimate that we will finish implementation faster. Release schedule should be discussed so that we release an official version with `range` support before you release a new version of b2_fuse.

Could you please do me a favor and file a feature request for download `range` parameter in b2 CLI issue tracker?

**Comment by falense** [comment]
I have added a feature request to the issue tracker. 

Best 
Sondre

**Comment by falense** [comment]
Hello again, 

I need a minimal example on how to create a bucket. As far as I can tell it seems I need to: 
1. Create an "Account info" object that is based on a "sqlite" database on disk?
2. Use said "Account info" object to create a B2API object
3. Use the B2API to create a Bucket object to do operations against.

Is this correct?

**Comment by ppolewicz** [comment]
You need an object that implements an [_AccountInfo_ interface](AbstractAccountInfo). It can be [_SqliteAccountInfo_](https://github.com/Backblaze/B2_Command_Line_Tool/blob/f41cfbba684234f3efd67e918d700ec707f34252/b2/account_info/sqlite_account_info.py#L26), but it could also be something that you keep only in memory and the user would feed the key and api key when mounting (that is how I imagined fuse will use it). B2 CLI does not currently have `InMemoryAccountInfo`, but it is easy to implement and it would be better if we owned it, than if you would (actually we have something like that for tests, so most of the work is done already).

As for creating a bucket, `test/test_bucket.py` is a good example. I have shortened it for you to this:

``` python
account_info = SqliteAccountInfo("/path/to/account-info.sqlite3")
api = B2Api(account_info)
api.authorize_account('production', 'account-id', 'api-key')
bucket = self.api.create_bucket('bucket_name', 'allPublic')
```

I'd advise against using `SqliteAccountInfo` though - you care about performance a lot and our implementation is might not be ideal. It cares about multi-process access using locks, journals, transactions etc, so a memory-based implementation seems to be much better for b2_fuse. If your understanding is the same, please confirm by creating a feature request for `InMemoryAccountInfo`.

**Comment by falense** [comment]
Sounds good. I have added another feature request: https://github.com/Backblaze/B2_Command_Line_Tool/issues/252

**Comment by mattrummler** [comment]
Hi as a beginning user of b2fuse AND backblaze b2 I may not have the perspective to provide much of value here.
(also thanks for b2fuse and for the b2 API :+1:  )

Having said that, it seems to me most of the caching issues could be resolved if the b2 API provided a way to get a list of files in a bucket that were changed since a certain timestamp/datetime.

Here is the process:
1- the b2 API client establishes a connection and pulls the full list of files from the bucket 
2- the client (or the b2 API service if possible?) stores the timestamp for the time the above list finishes
3- the client checks back when it sees fit & passes the timestamp to the b2 API
4- the b2 API returns only the files modified or added since the timestamp 
(NOTE: seems like the client would most effectively keep to a regularly timed cycle for the best performance... but would also obviously need to check back when the local client updates or adds a file)

Hope that helps (because I'd love to see the capability to list 1000s of files to multiple clients looking @ the same bucket, with a much lower level of latency)

**Comment by falense** [comment]
@mattrummler This is a bit off-topic but here it goes. I do not see any obvious flaws in such a design, but B2 buckets does not actually support timestamps (at least last time I checked). They do support arbitrary meta information in the form of key-value pairs, which is used to implement timestamps by for instance B2 Command line client. Since timestamps is not actually a feature supported by Backblaze it is a bit of leap to having them providing such an API feature. I agree it would be a useful one, as I have tried to (and failed) to find a good way to handle concurrent access without significant overhead and complexity.

**Comment by mattrummler** [comment]
@sondree Thanks for addressing my suggestion even though it's not directly related to the original issue. Was really trying to add some value to the discussion with @ppolewicz. 
Too bad, seems like timestamp capability would be extremely valuable for managing distributed change over large numbers of files in buckets.
@ppolewicz is there any chance that BackBlaze will consider adding the capability/is this a minimal effort addition or major??

Thank you both!

**Comment by ppolewicz** [comment]
B2 API is not finished. They are still working on extending it. B2 CLI team (as well as other b2 users) sometimes request changes to the API and sometimes Backblaze implements something based on user suggestions. Actually I think they rely heavily on user feedback to improve their service (as many smart companies do).

I, however, see a problem with the design that you suggested. Before we suggest it for implementation, let's assume for a moment that this is implemented. If I understood your design correctly, then the following could happen:

| server time | action |
| --- | --- |
| 1474935704.001 | Client-1: put a small file |
| 1474935705.101 | Server-1: put a file in sent in by Client-1, write down the time (**1474935705.101**), send a response |
| 1474935705.201 | Client-1: received a file upload confirmation |
| (~5 minutes pass) | meanwhile nothing happens on either side |
| 1474936005.001 | Client-1: getLastModifyTime/bucket-1 |
| 1474936005.002 | Client-2: put a small file |
| 1474936005.101 | Server-1 request received from Client-1, responding with 1474935705.101 ("5 minutes ago") |
| 1474936005.102 | Server-2: put a file sent in by Client-2, write down the time (**1474936005.102**!), send a response |
| 1474936005.201 | Client-1 received response: getLastModifyTime/bucket-1 -> **_1474935705.101_** -> that is the same time Client-1 had, so it judges that **bucket cache is up to date** |
| 1474936006.102 | Client-1: received a file upload confirmation |

In other words, Client can never assume that nothing has changed while the response for his request was in flight between Backblaze datacenter and itself.

This is a hard problem. @mattrummler you might want to read about Google Spanner - they tried to address a similar problem in a clever way in industrial conditions. You might also want to read about oplocks too.

To sum it up, I think they might opt to choose to implement a solution for the problem you mentioned, provided that:
- it is something that works reliably
- they'll judge that this will benefit their product more than other things they considered

**Comment by mattrummler** [comment]
@ppolewicz Good point. And this scenario increases in likelihood with an increase in activity & volume.
I will definitely go read about Spanner, sounds pretty interesting.
After reading through your analysis of the weakness in the previously suggested process, I can think of a reasonably easy set of changes to the process that would resolve this:
1. The timestamp that the client records should be current time minus a given amount of latency (this should be configurable based on the users tolerance for false positives vs. missing in-transit files...).
   While this is not a completely solid solution, you would get in-transit files on the following check, (so they would be delayed but they would show up, in the scenario you provided this would mean that the missed file would be caught the next time the client checked).
2. Next, the client would occasionally get a full list of files (basically renew it's list), to make 100% sure no files have been permanently missed. 
   (Also would be best to be a configurable option for a user so they get to choose accuracy vs. performance).
3. Finally, one more step to increase performance again, (though this one would be a lower priority in my opinion). 
   This part depends on BackBlaze providing the capability of assigning each uploaded file some unique ID**\* that represents uploads or changes by a specific client. 
   With this ID it would be possible for a client to differentiate between the files it uploaded vs. the files uploaded by others... This would allow the query to b2 to effectively ask the following

> "give me a list of all of the changes that have been made by other clients, which have occurred since ___ seconds before the last time I checked"

Anyway I suspect that would resolve the issues to the degree of accuracy 99.98% of users would care about. However, it all depends on b2 providing the capability of asking it to provide a list of files that have been uploaded or changed from a given time, (if it already does this and I missed something then I apologize but it was my understanding that they do not).

As a side note, I realize this modification, without step 3, would provide an updated list that includes files the current client uploaded. To me this is a low priority issue. The reason for this entire process is to prevent requiring the client to get the full list of files in a bucket just because they have made a change or the data is stale. This would take care of that problem. 

***(It sounds like this may already be possible utilizing the key value pair @sondree mentioned... assuming BackBlaze actually stores that key value pair on their end & that the client setup has a spot to add or create a unique ID[or really could just be the client's IP address if, as is my case, we can safely assume the clients are servers on the internet]...)

**Comment by ppolewicz** [comment]
Since you are going to read about Spanner and oplocks, lets pause this discussion until you do. 

I would not want to use a filesystem which is consistent **IF** latency did not spike. Would you? For me that _if_ is too big.

If a filesystem has inconsistent state, application that uses it can decide to do something terrible and it's developer is very unlikely to guarantee that it will be ok.

**Comment by ppolewicz** [comment]
@mattrummler also see this: https://en.m.wikipedia.org/wiki/Journaling_file_system
As b2 provides some atomicy already (incremental file version), we may be able to design a solution for consistent multi-access without any additional support in the Api. This is actually what I intended to do when I wanted to implement a b2 fuse myself. Later it turned out b2 CLI required a big effort to become usable as an underlying library for a filesystem and @sondre implemented something already, so I gave up on it. Synchronization of multiple clients was always something on my roadmap though.

**Comment by mattrummler** [comment]
@ppolewicz I appologize for my slowness. I did read up on Spanner & while they have a lot more technological capabilities available to them (many servers & developers etc...) part of their solution was to control the timestamps based on latency... the poor man's solution I came up with is very similar in effect...  (well to that part), it's just not dynamic and it's accuracy would be 100% after each step 2 of my revised process occurred, with the statistical likelihood of error increasing slightly to a maximum right before the next full refresh/step 2. 
Am reasonably familiar with journaling file systems because of past research into linux file systems, but I didn't even think of utilizing the change data they store! 
That seems like a really great idea. 
My use of b2fuse is in memory only, so it may not actually help in instances where there is no data being stored on the file system? 
(with my limited knowledge, I can't think of any way that it could unless fuse implementations can utilize the host file systems journal for data stored on the fuse mount...??)
Anyway thanks and hopefully backblaze with their larger resources, can quickly come up with something that will make all this a moot point.

**Comment by ppolewicz** [comment]
We can acquire consistent locks through remote journaling. Check which version "locks" file has, write a new version and check whether the one that we wrote has a version number immediately next to the one checked. If it is, then we have a lock. If not, then some other client has acquired the lock... So we wait until he releases it. A lock can have a long timeout, so that if a client crashes, filesystem is not permanently locked out. This is difficult, but possible. If two clients try to write at the same time, performance will be low, but hey, this is not how one should use b2fuse. If you want fast multi-access remote cloud storage with filesystem interface, it should use some other architecture for synchronization than b2 storage (which has high latency and syncs to 20 spinning hard drives on write...).

I don't like solutions which provide consistency with "high" probability - we should get it covered properly, so that it is always consistent unless there is a CPU error caused by overheating or something like that.

**Comment by falense** [comment]
Just a status update. Work is under way to convert B2Fuse to use B2 Command Line Tool as a backend. This will resolve this issue as B2 Command Line Tool gracefully handles retries and errors. See https://github.com/Backblaze/B2_Command_Line_Tool/issues/250 for details.

**Comment by falense** [comment]
Issue should now be resolved. I'll leave this open a bit longer in case it is not. Please give it a shot and report back if you still have troubles.

Best
